From: Kefeng Wang <wangkefeng.wang@huawei.com>
Subject: mm-memory-failure-make-action_result-return-int-v2
Date: Mon, 24 Oct 2022 11:51:38 +0800

- MF_DELAYED is considered as success, suggested by HORIGUCHI

- adjust order between unlock_page(head) and action_result()
  after !hwpoison_user_mappings() to clean code more in
  try_memory_failure_hugetlb()

Link: https://lkml.kernel.org/r/20221024035138.99119-1-wangkefeng.wang@huawei.com
Signed-off-by: Kefeng Wang <wangkefeng.wang@huawei.com>
Acked-by: Naoya Horiguchi <naoya.horiguchi@nec.com>
Cc: Miaohe Lin <linmiaohe@huawei.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/memory-failure.c |    9 +++------
 1 file changed, 3 insertions(+), 6 deletions(-)

--- a/mm/memory-failure.c~mm-memory-failure-make-action_result-return-int-v2
+++ a/mm/memory-failure.c
@@ -1191,7 +1191,7 @@ static int action_result(unsigned long p
 	pr_err("%#lx: recovery action for %s: %s\n",
 		pfn, action_page_types[type], action_name[result]);
 
-	return result == MF_RECOVERED ? 0 : -EBUSY;
+	return (result == MF_RECOVERED || result == MF_DELAYED) ? 0 : -EBUSY;
 }
 
 static int page_action(struct page_state *ps, struct page *p,
@@ -1202,14 +1202,12 @@ static int page_action(struct page_state
 	/* page p should be unlocked after returning from ps->action().  */
 	result = ps->action(ps, p);
 
-	action_result(pfn, ps->type, result);
-
 	/* Could do more checks here if page looks ok */
 	/*
 	 * Could adjust zone counters here to correct for the missing page.
 	 */
 
-	return (result == MF_RECOVERED || result == MF_DELAYED) ? 0 : -EBUSY;
+	return action_result(pfn, ps->type, result);
 }
 
 static inline bool PageHWPoisonTakenOff(struct page *page)
@@ -1890,9 +1888,8 @@ retry:
 	page_flags = head->flags;
 
 	if (!hwpoison_user_mappings(p, pfn, flags, head)) {
-		res = action_result(pfn, MF_MSG_UNMAP_FAILED, MF_IGNORED);
 		unlock_page(head);
-		return res;
+		return action_result(pfn, MF_MSG_UNMAP_FAILED, MF_IGNORED);
 	}
 
 	return identify_page_state(pfn, p, page_flags);
_
